### PR TITLE
Iteration api by section

### PR DESCRIPTION
**Pull request by mugdhadhole1**
#12

### CODE REVIEW

File: trlc/ast.py [hunk 1]
```diff
@@ -3082,6 +3084,24 @@ def all_names(self):
...
+            if record_object.section:
+                self.section_names = record_object.section
+                if len(self.section_names) > 0:
+                    for level in range(len(self.section_names)):
+                        if self.section_names[level].name not in section_dictionary.keys():
+                            section_dictionary.update({self.section_names[level].name: True})                            
+                            yield self.section_names[level].name, level
+                    yield record_object.name
+            else:
+                yield record_object.name
```
Comment (christophkloeffel) [inline trlc/ast.py:3103]: a proposal for refactoring the code:
get rid of self.section_dictionary then. not necessary
```suggestion
        object_level = len(record_object.section)-1 if record_object.section else 0
        if record_object.section:
            for level, section in enumerate(record_object.section):
                if section.name not in section_names:
                    section_names.append(section.name)
                    yield section.name, level
            yield record_object, object_level
        else:
            yield record_object, object_level
```

File: trlc/ast.py [hunk 2]
```diff
@@ -3082,6 +3084,24 @@ def all_names(self):
             rv |= self.parent.all_names()
         return rv
 
+    def iter_record_objects_by_section(self):
```
Comment (christophkloeffel) [inline trlc/ast.py:3087]: add at least some kind of docstring which explains what this method is yielding

**Review by christophkloeffel — CHANGES_REQUESTED** [review]

File: trlc/parser.py [hunk 3]
```diff
@@ -1778,7 +1776,7 @@ def parse_record_object_declaration(self):
             name      = self.ct.value,
             location  = self.ct.location,
             n_typ     = r_typ,
-            section   = self.section[-1] if self.section else None,
+            section   = self.section if self.section else None,
```
Comment (christophkloeffel) [inline trlc/parser.py:1779]: is this even working without `self.section.copy()`?

File: trlc/parser.py [hunk 4]
```diff
@@ -1535,14 +1535,12 @@ def parse_section_declaration(self):
...
-        if self.section:
-            sec = ast.Section(name     = self.ct.value,
-                              location = self.ct.location,
-                              parent   = self.section[-1])
-        else:
-            sec = ast.Section(name     = self.ct.value,
-                              location = self.ct.location,
-                              parent   = None)
+
+        self.match_kw("section")
```
[the remark this answers is not in the document]
Comment (florianschanda) [reply trlc/parser.py:1539]: Yeah I am pretty sure this is wrong :) Copy-pasta error?

**Review by christophkloeffel — APPROVED** [review]

**Pull request merged**